runtime-sdk/src/state.rs

```diff
@@ -0,0 +1,1545 @@
+            store.commit();
+        });
+    }
+}
+
+struct TransactionGuard(usize);
+
+impl Drop for TransactionGuard {
+    fn drop(&mut self) {
+        let level = CurrentState::with(|state| state.level());
+
+        // If transaction hasn't been either committed or reverted, rollback.
+        if level == self.0 {
+            CurrentState::rollback_transaction();
+        }
+    }
+}
+
+/// Result of a transaction helper closure.
+pub enum TransactionResult<T> {
+    Commit(T),
+    Rollback(T),
+}
+
+impl From<()> for TransactionResult<()> {
+    fn from(_: ()) -> TransactionResult<()> {
+        TransactionResult::Commit(())
+    }
+}
+
+impl<R, E> From<Result<R, E>> for TransactionResult<Result<R, E>> {
+    fn from(v: Result<R, E>) -> TransactionResult<Result<R, E>> {
+        match v {
+            Ok(_) => TransactionResult::Commit(v),
+            Err(_) => TransactionResult::Rollback(v),
+        }
+    }
+}
+
+/// State attached to the current thread.
+pub struct CurrentState;
+
+impl CurrentState {
+    /// Attach a new state to the current thread and enter the state's context.
+    ///
+    /// The passed store is used as the root store.
+    pub fn enter<S, F, R>(root: S, f: F) -> R
+    where
+        S: Store,
+        F: FnOnce() -> R,
+    {
+        Self::enter_opts(Options::default(), root, f)
+    }
+
+    /// Attach a new state to the current thread and enter the state's context.
+    ///
+    /// The passed store is used as the root store.
+    pub fn enter_opts<S, F, R>(opts: Options, mut root: S, f: F) -> R
+    where
+        S: Store,
+        F: FnOnce() -> R,
+    {
+        let root = unsafe {
+            // SAFETY: Keeping the root store is safe as it can only be accessed from the current
+            // thread while we are running inside `CurrentState::enter` where we are holding a
+            // mutable reference on it.
+            std::mem::transmute::<_, &mut (dyn Store + 'static)>(&mut root as &mut dyn Store)
+        };
+        // Initialize the root state.
+        let mut root = State {
+            parent: None,
+            store: Some(OverlayStore::new(Box::new(root) as Box<dyn Store>)),
+            events: EventTags::new(),
+            unconditional_events: EventTags::new(),
+            messages: Vec::new(),
+            block_values: BTreeMap::new(),
+            hidden_block_values: None,
+            local_values: BTreeMap::new(),
+            rng: Some(RootRng::new()),
+            hidden_rng: None,
+            env: Default::default(),
+            always_rollback: false,
+        };
+        // Apply options to allow customization of the root state.
+        opts.apply(&mut root);
+
+        CURRENT.with(|c| {
+            c.try_borrow_mut()
+                .expect("must not re-enter from with block")
+                .push(root)
+        });
+        let _guard = CurrentStateGuard; // Ensure current state is popped once we return.
```

Comment:
I usually see these things with a symmetry to them, such that creating one is combined with pushing the state

Reply:
Yeah, it is pushed just above.